src/evm.rs

```diff
@@ -1,179 +1,338 @@
+use RevmExecutionResult::Success;
+use tracing::trace;
+
+use crate::{types::{AccountInfo, Env, ExecutionResult, JournalCheckpoint}, utils::{addr, pyerr}};
+use crate::database::DB;
+use crate::executor::call_evm;
+use crate::types::{PyByteVec, PyDB};
+
+#[derive(Debug)]
 #[pyclass]
-pub struct EVM(Executor);
+pub struct EVM {
+    /// Context of execution.
+    context: EvmContext<DB>,
 
-impl EVM {
-    pub fn db(&self) -> &Backend {
-        &self.0.backend
-    }
-}
+    /// Handler configuration.
+    handler_cfg: HandlerCfg,
 
-fn pyerr<T: Debug>(err: T) -> pyo3::PyErr {
-    PyRuntimeError::new_err(format!("{:?}", err))
+    /// The gas limit for calls and deployments. This is different from the gas limit imposed by
+    /// the passed in environment, as those limits are used by the EVM for certain opcodes like
+    /// `gaslimit`.
+    gas_limit: U256,
+
+    /// whether to trace the execution to stdout
+    tracing: bool,
+
+    /// Checkpoints for reverting state
+    /// We cannot use Revm's checkpointing mechanism as it is not serializable
+    checkpoints: HashMap<JournalCheckpoint, RevmCheckpoint>,
+
+    /// The result of the last transaction
+    result: Option<RevmExecutionResult>,
 }
 
 #[pymethods]
 impl EVM {
+    /// Create a new EVM instance.
     #[new]
-    #[pyo3(signature = (env=None, fork_url=None, fork_block_number=None, gas_limit=18446744073709551615, tracing=false))]
+    #[pyo3(signature = (env = None, fork_url = None, fork_block_number = None, gas_limit = 18446744073709551615, tracing = false, spec_id = "LATEST"))]
     fn new(
         env: Option<Env>,
-        fork_url: Option<String>,
-        fork_block_number: Option<u64>,
+        fork_url: Option<&str>,
+        fork_block_number: Option<&str>,
         gas_limit: u64,
         tracing: bool,
+        spec_id: &str,
     ) -> PyResult<Self> {
-        let evm_opts = EvmOpts {
-            fork_url: fork_url.clone(),
-            fork_block_number,
-            ..Default::default()
+        let spec = SpecId::from(spec_id);
+        let env = env.unwrap_or_default().into();
+        let db = fork_url.map(|url| DB::new_fork(url, fork_block_number)).unwrap_or(Ok(DB::new_memory()))?;
+
+        let Evm { context, .. } = Evm::builder()
+            .with_env(Box::new(env))
+            .with_db(db)
+            .build();
+        Ok(EVM {
+            context: context.evm,
+            gas_limit: U256::from(gas_limit),
+            handler_cfg: HandlerCfg::new(spec),
+            tracing,
+            checkpoints: HashMap::new(),
+            result: None,
+        })
+    }
+
+    fn snapshot(&mut self) -> PyResult<JournalCheckpoint> {
+        let checkpoint = JournalCheckpoint {
+            log_i: self.context.journaled_state.logs.len(),
+            journal_i: self.context.journaled_state.journal.len(),
         };
+        self.checkpoints.insert(checkpoint, self.context.journaled_state.checkpoint());
```

Comment:
Checkpoint increments depth, this should be decoupled in Revm. and move depth from `JournaledState` to `EvmContext`

Reply:
That would be great, but that's a whole separate task